### PR TITLE
improve coverage

### LINKED ISSUE #195: improve test coverage
These functions have relatively low test coverage. 

- [`formatting_reply` and `on_type_formatting_reply`](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/formatting.R)
- [`document_highlight_reply`](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/highlight.R)
- [`hover_reply`](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/hover.R)
- [`document_link_reply `](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/link.R)
- [`document_symbol_reply`](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/symbol.R)

### PR DESCRIPTION
**Pull request by randy3k**
close #195 

- [x] [`completion_reply`](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/completion.R)
- [x] [`formatting_reply` and `on_type_formatting_reply`](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/formatting.R)
- [x] [`document_highlight_reply`](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/highlight.R)
- [x] [`hover_reply`](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/hover.R)
- [x] [`document_link_reply `](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/link.R)
- [x] [`document_symbol_reply`](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/symbol.R)
- [x] [`document_color_reply`](https://codecov.io/gh/REditorSupport/languageserver/src/master/R/color.R)

### CODE REVIEW

**Comment by renkun-ken** [comment]
I'm also interested in writing the test cases. Would you mind I push to this branch directly?

**Comment by randy3k** [comment]
No problem. I don’t have time to work on it lately.

**Comment by renkun-ken** [comment]
I tried adding test cases for code section symbols but I'm not sure where to specify client capabilities to make the testing language client to support `hierarchicalDocumentSymbolSupport`. I tried the following 

```r
    client <- language_client()
    client$ClientCapabilities <- list(
        textDocument = list(
            documentSymbol = list(
                hierarchicalDocumentSymbolSupport = TRUE
            )
        )
    )
```

but it does not seem to work as `document_symbol_reply` still gets `self$ClientCapabilities = NULL`.

**Comment by randy3k** [comment]
The information for `ClientCapabilities` is sent when `language_client` is execulated. You will need to set `ClientCapabilities` in the `initialize` function at https://github.com/REditorSupport/languageserver/blob/38122caaf3ecc1e914294c6010887391da2c3b53/R/languageclient.R#L20-L20

**Comment by randy3k** [comment]
My bad, the code was already there https://github.com/REditorSupport/languageserver/blob/38122caaf3ecc1e914294c6010887391da2c3b53/R/languageclient.R#L104-L104

**Comment by renkun-ken** [comment]
Thanks! Now new test cases are added to all providers listed.

**Comment by randy3k** [comment]
Wow, amazing job. With the tests, I guess we are already for the next release.

**Comment by renkun-ken** [comment]
I'm also trying to add test cases for scope completion but I'm not sure why the following tests are not working:

```r
test_that("Completion of symbols in scope works", {
    skip_on_cran()
    client <- language_client()

    withr::local_tempfile(c("temp_file"), fileext = ".R")
    writeLines(
        c(
            "my_fun <- function(test1) {",
            "    test2 <- 1",
            "    local({",
            "        for (test3 in 1:10) {",
            "            test",
            "        }",
            "    })",
            "}"
        ),
        temp_file
    )

    client %>% did_save(temp_file)

    result <- client %>% respond_completion(
        temp_file, c(4, 16),
        retry_when = function(result) length(result) == 0 || length(result$items) == 0
    )

    print(str(result))
})
```

It does not print any locally scoped symbols:

```
List of 2
 $ isIncomplete: logi FALSE
 $ items       :List of 3
  ..$ :List of 3
  .. ..$ label: chr "testthat"
  .. ..$ kind : int 9
  .. ..$ data :List of 1
  .. .. ..$ type: chr "package"
  ..$ :List of 6
  .. ..$ label           : chr "testInheritedMethods"
  .. ..$ kind            : int 3
  .. ..$ detail          : chr "{methods}"
  .. ..$ insertText      : chr "testInheritedMethods($0)"
  .. ..$ insertTextFormat: int 2
  .. ..$ data            :List of 2
  .. .. ..$ type   : chr "function"
  .. .. ..$ package: chr "methods"
  ..$ :List of 6
  .. ..$ label           : chr "testVirtual"
  .. ..$ kind            : int 3
  .. ..$ detail          : chr "{methods}"
  .. ..$ insertText      : chr "testVirtual($0)"
  .. ..$ insertTextFormat: int 2
  .. ..$ data            :List of 2
  .. .. ..$ type   : chr "function"
  .. .. ..$ package: chr "methods"
```

Looks like the file is not parsed?

**Comment by renkun-ken** [comment]
Test cases for Rmarkdown are divided into handler test scripts and many test cases now have their Rmd version added.

**Comment by randy3k** [comment]
👍 

By the way, I don't understand why travis/push fails but travis/pr does not. Each time it fails at the same test.

File: tests/testthat/test-formatting.R [hunk 1]
```diff
@@ -40,14 +54,79 @@ test_that("Formatting selection works for partial line", {
...
+    expect_length(result, 1)
+    lines <- strsplit(result[[1]]$newText, "\n")[[1]]
+    expect_length(lines, 4)
+    expect_equal(lines, c(
+        "my_fn <- function(x) {",
+        "    f(x + 1, x - 1)",
+        "    data[x, y]",
+        "}"
+    ))
+})
```
Comment (randy3k) [inline tests/testthat/test-formatting.R]: This test fails on travis/pr as well as my local computer. I don't understand how we get the indented `f` and `data`.

I keep getting
```
handling request:  textDocument/onTypeFormatting
formatting error: <text>:1:1: unexpected '}'
1: }
    ^
```

Comment (randy3k) [reply tests/testthat/test-formatting.R]: I also don't understand how github actions tests pass...

Comment (randy3k) [reply tests/testthat/test-formatting.R]: Oh, actually the tests in github actions actually failed, but the badge is not correct.

Comment (randy3k) [reply tests/testthat/test-formatting.R]: 8f0bdcf fixed the github actions tests

Comment (renkun-ken) [reply tests/testthat/test-formatting.R]: Just notice that the `coverage` branch is behind #209. We may rebase to latest master to resolve this?

Comment (randy3k) [reply tests/testthat/test-formatting.R]: I have rebased the branch onto master.

Comment (randy3k) [reply tests/testthat/test-formatting.R]: Good catch @renkun-ken. That also explains why travis/push failed but travis/pr didn't

**Comment by renkun-ken** [comment]
Looks like some code is reverted? 

```r
language_client <- function(working_dir = getwd(), debug = FALSE, diagnostics = FALSE)
```

where the capabilities argument is gone?

**Comment by randy3k** [comment]
> Looks like some code is reverted?
> 
> ```r
> language_client <- function(working_dir = getwd(), debug = FALSE, diagnostics = FALSE)
> ```
> 
> where the capabilities argument is gone?

I guess I have accidentally removed the diff when I was resolving the conflicts.

**Comment by randy3k** [comment]
Looks like there is a path related issue in Document link provider on Windows.

**Comment by renkun-ken** [comment]
Can't see the error details except for those expect lines. Do we have a way to know more details?

**Comment by randy3k** [comment]
Perhaps we could always enable the debug setting https://github.com/REditorSupport/languageserver/blob/master/tests/testthat/helper-utils.R#L14
and dump the file if there are any errors.

**Comment by renkun-ken** [comment]
Maybe I should install a Windows virtual machine to have a look at this. 😂

**Comment by randy3k** [comment]
@renkun-ken see if the dumped log for windows would help
https://github.com/REditorSupport/languageserver/actions/runs/41699817

**Comment by randy3k** [comment]
It doesn't have much information in the log as `document_link_reply` doesn't have any debug log.

**Comment by renkun-ken** [comment]
I've set up a Windows virtual box and am looking into this.😂

**Comment by renkun-ken** [comment]
It looks like there's something inconsistent in slash between path and uris. According to https://microsoft.github.io/language-server-protocol/specifications/specification-current/#uri, uri should only use `/`. However, under Windows, `normalizePath()` uses `\\` by default while `file.path()` uses `/`, and `path_to_uri` does not ensure `/` so the resulted URI may become something like

```
"file:///C:%5CUsers%5CRUNNER~1%5CAppData%5CLocal%5CTemp%5CRtmpakDrf8/working_dir%5CRtmpC2WiHu%5Cfile16401a762f8e.R"
```

if the input path uses `\\` since it is URL encoded.

**Comment by renkun-ken** [comment]
On my Windows 10 VM, I tried to run the tests but link test always timeout.

**Comment by renkun-ken** [comment]
After some painful debugging on Windows 10 VM, the cause it found:

The link test cases write the following code to the temp file:

```r
source('C:\Users\ken\AppData\Local\Temp\Rtmp6HblZk\file1b5448882fd4.R')
source('file1b5448882fd4.R')
source('non_exist_file.R')
```

The Windows slash in the file makes it not parse-able.

**Pull request merged**